File: app/src/at-sonnguyen/java/com/asiantech/intern20summer1/Intro/EdgeOfTheOcean.java
package com.asiantech.intern20summer1.Intro;

public class EdgeOfTheOcean {
    public static void main(){
    int Shape = shapeArea(15);
    int [ ] inputArray = {1,2,5,8,9};
    int AddToArray = makeArrayConsecutive2(inputArray);
    boolean checkSequence = almostIncreasingSequence(inputArray);
    int [][] Matrix = {{2,3,5},{1,3,6},{5,4,9}};
    int Sum = matrixElementsSum(Matrix);
    }
    /*
    * Given an array of integers, find the pair of adjacent elements that has the largest product and return that product.
    * */
    static int adjacentElementsProduct(int[] inputArray) {
        int max = inputArray[0] * inputArray[1];
        for (int i = 0; i < inputArray.length - 1; i++) {
            if (inputArray[i] * inputArray[i + 1] > max) {
                max = inputArray[i] * inputArray[i + 1];
            }
        }
        return max;
    }
    /*
    *Below we will define an n-interesting polygon.
    * Your task is to find the area of a polygon for a given n.
    * An n-interesting polygon is obtained by taking the n - 1-interesting
    polygon and appending 1-interesting polygons to its rim, side by side.
    * You can see the 1-, 2-, 3- and 4-interesting polygons in the picture below.
    */
    static int shapeArea(int n) {
        return n*n + (n-1)*(n-1);
    }
    /*
    * Ratiorg got statues of different sizes as a present from CodeMaster for his birthday,
    * each statue having an non-negative integer size. Since he likes to make things perfect,
    * he wants to arrange them from smallest to largest so that each statue will be bigger
     than the previous one exactly by 1.
    * He may need some additional statues to be able to accomplish that.
    * Help him figure out the minimum number of additional statues needed.
    * */
    static int makeArrayConsecutive2(int[] statues) {
        int max = statues[0];
        int min = statues[0];
        for (int i = 0;i<statues.length; i++){

            if (statues[i]> max){
                max = statues[i];
            }
            if (statues[i]<min){
                min = statues[i];
            }
        }
        return (max - min - statues.length + 1);
    }
    /*
    * Given a sequence of integers as an array,
    * determine whether it is possible to obtain a strictly increasing sequence
    * by removing no more than one element from the array.
    * */
    static boolean almostIncreasingSequence(int[] sequence) {
        int count1 = 0, count2 = 0;
        for (int i = 0; i < sequence.length-1;i++){
            if (sequence[i]>=sequence[i+1]){
                count1 +=1;
            }
        }
        for (int i = 0; i < sequence.length-2;i++){
            if (sequence[i]>=sequence[i+2]){
                count2 +=1;
            }
        }
        if (count1 > 1  ) {
            return false;
        }else if (count2 >1){
            return false;
        }else return true;

    }
    /*
    * Given matrix, a rectangular matrix of integers,
    *  where each value represents the cost of the room,
    * your task is to return the total sum of all rooms that are suitable for the CodeBots
    * */
    static int matrixElementsSum(int[][] matrix) {
        int Sum =0;
        for(int i=0;i<matrix.length;i++){
            for(int j=0;j<matrix[0].length;j++){
                if(matrix[i][j]==0){
                    for(int k=i;k<matrix.length;k++){
                        matrix[k][j]=0;
                    }
                    continue;
                }
                else
                    Sum+=matrix[i][j];
            }
        }
        return Sum;
    }
}

File: app/src/at-sonnguyen/java/com/asiantech/intern20summer1/kotlin/EdgeOfTheOcean.kt
package com.asiantech.intern20summer1.kotlin

object EdgeOfTheOcean {
    const val SHAPE = 5

    /*
    *   Given an array of integers,
    *   find the pair of adjacent elements that has the largest product and return that product
    */
    private fun adjacentElementsProduct(inputArray: MutableList<Int>): Int {
        var lengthOfArray = inputArray.size
        var max : Int = inputArray[0]*inputArray[1]
        for (index in 0 until lengthOfArray-1){
            if(inputArray[index]*inputArray[index+1] > max){
                max = inputArray[index]*inputArray[index+1]
            }
        }
        return max
    }
    /*
    *   return the area of a polygon for a given n
    */
    private fun shapeArea(n: Int): Int = n*n + (n-1)*(n-1)
    /*
    *   the minimum number of additional statues needed.
    */
    private fun makeArrayConsecutive2(statues: MutableList<Int>): Int {
        statues.sort()
        return statues[statues.size-1] - statues[0] - statues.size +1
    }
    /*
    *   Given a sequence of integers as an array,
    * determine whether it is possible to obtain a strictly increasing sequence
    * by removing no more than one element from the array
    */
    private fun almostIncreasingSequence(sequence: MutableList<Int>): Boolean {
        var errorCountOne = 0
        var errorCountTwo = 0

        for(i in 0..sequence.size - 2)
        {
            if(sequence[i] >= sequence[i + 1])
                errorCountOne++

            if(i != 0)
            {
                if(sequence[i - 1] >= sequence[i + 1])
                    errorCountTwo++
            }
        }

        return errorCountOne == 1 && errorCountTwo <= 1
    }
    /*
    * return the total sum of all rooms that are suitable for the CodeBots
    */
    private fun matrixElementsSum(matrix: MutableList<MutableList<Int>>): Int {
        var sum : Int = 0
        for(i in 0 until matrix.size){
            for (j in 0 until matrix[0].size){
                if (matrix[i][j]==0){
                    for (k in i until matrix.size){
                        matrix[k][j] =0
                    }
                    continue
                }else sum += matrix[i][j]

            }
        }
        return sum
    }

    @JvmStatic
    fun main(args: Array<String>) {
        var inputMatrix : MutableList<MutableList<Int>> = mutableListOf(mutableListOf(0,1,1,2), mutableListOf(0,5,0,0),
            mutableListOf(2,0,3,3) )
        var inputArray : MutableList<Int> = mutableListOf(3, 6, -2, -5, 7, 3)
        var n = SHAPE
        println("4. ${adjacentElementsProduct(inputArray)}")
        println("5. Shape Area of $n = ${shapeArea(n)}")
        println("6. minimum add number = ${makeArrayConsecutive2(inputArray)} ")
        println("7. $inputArray is almost increasing Array ? ${almostIncreasingSequence(inputArray)}")
        println("8. max of Element Sum in $inputMatrix = ${matrixElementsSum(inputMatrix)}")
    }
}
